File: sources/scala/tools/scalac/transformer/matching/TracerInScala.scala
import scalac.ast.Tree;
import scalac.symtab.Symbol;
import scalac.symtab.Type;
//import scalac.transformer.matching.CodeFactory ;
import scalac.transformer.matching.DetWordAutom ;

package scala.tools.scalac.transformer.matching {
/** @todo: factor common things of LeftTracerInScala and RightTracerInScala
 */
class TracerInScala(dfa: DetWordAutom, elementType: Type, owner: Symbol, cf: CodeFactory )
extends Autom2Scala(dfa, elementType, owner, cf) {

  override var optimize = true;

  final def collectVars(pat: Tree) = {
    val c = new CollectVariableTraverser();
    c.collectVars(pat)
  }

  final def containsBinding(pat: Tree) = {
    val c = new CollectVariableTraverser();
    c.containsBinding(pat);
  }

}
}

File: sources/scalac/transformer/matching/TracerInScala.java
package scalac.transformer.matching ;

import scalac.*;
import scalac.ast.*;
import scalac.symtab.*;
import Tree.*;
import scalac.util.Name;
import scalac.util.Names;

import scalac.transformer.TransMatch.Matcher ;

import java.util.* ;

import ch.epfl.lamp.util.Position;


/** 2do: factor common things of LeftTracerInScala and RightTracerInScala
 */
public class TracerInScala extends Autom2Scala {


    HashMap helpMap;

      public TracerInScala( DetWordAutom dfa,
			    Type elementType,
			    Symbol ownerSym,
			    CodeFactory cf ) {
	super( dfa, elementType, ownerSym, cf );

	this.helpMap = new HashMap();

      }

      Tree refHelpVar( Symbol realVar ) {
            Symbol hv = (Symbol)helpMap.get( realVar );
            assert hv != null : realVar;
            return gen.Ident(Position.FIRSTPOS, hv);
      }

      Tree assignToHelpVar( Symbol realVar, Tree rhs ) {
            Tree hv = refHelpVar( realVar );
            return gen.Assign( hv, rhs );
      }

      Tree bindVar(Symbol realVar) {
            Tree hv = refHelpVar( realVar );
            //System.out.println("binding realVar.name "+realVar.name+" type:"+realVar.type()+" to smth");
            realVar.setOwner( owner );
            if( realVar.type().isSameAs( elementType ))
                  return gen.ValDef( realVar, cf.SeqList_head( hv ));
            else
                  return gen.ValDef( realVar, hv);

      }

}

